rust/fury-core/src/resolver/class_resolver.rs

```diff
@@ -18,99 +18,78 @@
+
+fn serialize<T: 'static + Serializer>(this: &dyn Any, context: &mut WriteContext) {
+    let this = this.downcast_ref::<T>().unwrap();
+    T::serialize(this, context)
 }
 
 impl ClassInfo {
-    pub fn new<T: StructSerializer>(fury: &Fury, type_id: u32) -> ClassInfo {
+    pub fn new<T: Serializer>(fury: &Fury, type_id: u32) -> ClassInfo {
         ClassInfo {
             type_def: T::type_def(fury),
-            type_id,
+            fury_type_id: type_id,
+            rust_type_id: TypeId::of::<T>(),
+            trait_object_serializer: serialize::<T>,
+            trait_object_deserializer: T::get_trait_object_deserializer(),
         }
     }
 
-    pub fn get_type_id(&self) -> u32 {
-        self.type_id
+    pub fn get_rust_type_id(&self) -> TypeId {
+        self.rust_type_id
+    }
+
+    pub fn get_fury_type_id(&self) -> u32 {
+        self.fury_type_id
     }
 
     pub fn get_type_def(&self) -> &Vec<u8> {
         &self.type_def
     }
+
+    pub fn get_serializer(&self) -> fn(&dyn Any, &mut WriteContext) {
+        self.trait_object_serializer
+    }
+
+    pub fn get_trait_object_deserializer<T: 'static>(&self) -> Option<&TraitObjectDeserializer> {
+        let type_id = TypeId::of::<T>();
+        self.trait_object_deserializer.get(&type_id)
+    }
 }
 
 #[derive(Default)]
 pub struct ClassResolver {
-    serialize_map: HashMap<u32, Harness>,
-    type_id_map: HashMap<TypeId, u32>,
+    fury_type_id_map: HashMap<u32, TypeId>,
     class_info_map: HashMap<TypeId, ClassInfo>,
 }
 
 impl ClassResolver {
-    pub fn get_class_info(&self, type_id: TypeId) -> &ClassInfo {
+    pub fn get_class_info_by_rust_type(&self, type_id: TypeId) -> &ClassInfo {
         self.class_info_map.get(&type_id).unwrap()
     }
 
-    pub fn register<T: StructSerializer>(&mut self, class_info: ClassInfo, id: u32) {
-        fn serializer<T2: 'static + StructSerializer>(this: &dyn Any, context: &mut WriteContext) {
-            let this = this.downcast_ref::<T2>();
-            match this {
-                Some(v) => {
-                    T2::serialize(v, context);
-                }
-                None => todo!(),
-            }
-        }
-
-        fn deserializer<T2: 'static + StructSerializer>(
-            context: &mut ReadContext,
-        ) -> Result<Box<dyn Any>, Error> {
-            match T2::deserialize(context) {
-                Ok(v) => Ok(Box::new(v)),
-                Err(e) => Err(e),
-            }
-        }
-        self.type_id_map.insert(TypeId::of::<T>(), id);
-        self.serialize_map
-            .insert(id, Harness::new(serializer::<T>, deserializer::<T>));
-        self.class_info_map.insert(TypeId::of::<T>(), class_info);
-    }
-
-    pub fn get_harness_by_type(&self, type_id: TypeId) -> Option<&Harness> {
-        self.get_harness(*self.type_id_map.get(&type_id).unwrap())
+    pub fn get_class_info_by_fury_type(&self, type_id: u32) -> &ClassInfo {
+        let type_id = self.fury_type_id_map.get(&type_id).unwrap();
```

Comment:
Too many unwrap here, how about return a Result?

Reply:
Yes. The codebase contain too many unwrap, I will fix it one by one.